docs/debugging.md

```diff
@@ -271,4 +271,90 @@ slower than a compiled version, so you should default to installing the wheel in
 CI instead of compiling Cython, which can take up to a few minutes on some CI
 runners.
 
+### Compiling CPython and foundational packages with thread sanitizer (TSAN)
+
+[Thread sanitizer](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual) (or TSAN) helps
+to detect C/C++ data races in concurrent systems. This tool can help to reveal free-threading
+related bugs in CPython and foundational packages (e.g. `numpy`).
+In this section we provide the commands to build free-threading compatible CPython and packages with TSAN and other hints to discover potential data races.
+
+#### Compile free-threading CPython with TSAN
+
+- Clone free-threading stable branch, e.g. 3.13.1
+
+```bash
+git clone https://github.com/python/cpython.git -b v3.13.1
+```
+
+- Configure and build (we use clang-18 as compiler):
+
+```bash
+cd cpython
+CC=clang-18 CXX=clang++-18 ./configure --disable-gil --with-thread-sanitizer --prefix $PWD/cpython-tsan
+make -j 8
+make install
+```
+
+- Use built python interpreter:
+
+```bash
+export PATH=$PWD/cpython-tsan/bin:$PATH
+python3.13t -VV
+# Python 3.13.1 experimental free-threading build (tags/v3.13.1:06714517797, Dec 19 2024, 10:06:54) [Clang 18.1.3 (1ubuntu1)]
+PYTHON_GIL=0 python3.13t -c "import sys; print(sys._is_gil_enabled())"
+# False
+```
+
+#### Compile NumPy with TSAN
+
+- Get the source code (for example, `main` branch)
+
+```bash
+git clone --recursive https://github.com/numpy/numpy.git
+```
+
+- Install build requirements:
+
+```bash
+cd numpy
+python3.13t -mpip install -r requirements/build_requirements.txt
+# Make sur to install compatible Cython version
```

Comment:
typo in "sure". more importantly: could you change to also install Cython from its `master` branch? There's at least one critical bug for fused types fixed after `3.1.0a1`.

Reply:
How about this?
```
python -m pip install -U git+https://github.com/cython/cython
```